### PR TITLE
Multi-threaded RPC server in SPRIGHT Gateway

### PR DESCRIPTION
**Pull request by sungyu-chang**
This PR splits the RPC server into two separate threads: a setup thread and an RX thread. The RX thread uses epoll for receiving. The setup thread is responsible for adding client sockets to epoll.

This PR also fixed the routing bug when the direct function routing (DFR) mode is disabled.

### CODE REVIEW

**Comment by ShixiongQi** [comment]
Need to use a better name to describe PR

File: src/gateway.c [hunk 1]
```diff
@@ -94,8 +94,17 @@ static void configure_keepalive(int sockfd) {
         exit(EXIT_FAILURE);
     }
 }
-
-static int get_client_info(int client_socket) {
+/**
```
Comment (ShixiongQi) [inline src/gateway.c:98]: Add an empty line before this line.

File: src/gateway.c [hunk 2]
```diff
@@ -207,81 +222,116 @@ static int rpc_server(void) {
...
+error:
+    ret = close(sockfd_l);
+    // TODO: close the epoll fd gracefully
+    if (unlikely(ret == -1)) {
+        log_error("close() error: %s", strerror(errno));
+        return -1;
     }
+    return 0;
+}
+static int rpc_server_receive(int epfd) {
```
Comment (ShixiongQi) [inline src/gateway.c:258]: Add an empty line before this line.

File: src/gateway.c [hunk 3]
```diff
@@ -207,81 +222,116 @@ static int rpc_server(void) {
 
     if (cfg->n_nodes == 1) {
         log_warn("No PEER NODE CONFIGURED. Terminating the RPC server...");
-        goto error_2;
+        goto error;
     }
-
-    for (i = 0; i < cfg->n_nodes - 1; i++) {
+    char client_addr[INET_ADDRSTRLEN];
```
Comment (ShixiongQi) [inline src/gateway.c]: Add an empty line before this line.

File: src/gateway.c [hunk 4]
```diff
@@ -207,81 +222,116 @@ static int rpc_server(void) {
...
+    while (1) {
         sockfd_c = accept(sockfd_l, NULL, NULL);
         if (unlikely(sockfd_c == -1)) {
             log_error("accept() error: %s",
                     strerror(errno));
-            return -1;
+            goto error;
         }
-
+        get_client_info(sockfd_c, client_addr, INET_ADDRSTRLEN);
```
Comment (ShixiongQi) [inline src/gateway.c]: Add an empty line before this line.

File: src/gateway.c (hunk 3 above)
Comment (ShixiongQi) [inline src/gateway.c]: Where is "client_addr" used? Please remove "client_addr" if it is not used elsewhere.

File: src/gateway.c [hunk 5]
```diff
@@ -207,81 +222,116 @@ static int rpc_server(void) {
...
-            log_error("rte_mempool_get() error: %s",
-                    rte_strerror(-ret));
-            goto error_0;
+        n_events = epoll_wait(epfd, event, N_EVENTS_MAX, -1);
+        if (unlikely(n_events == -1)) {
+            log_error("epoll_wait() error: %s",
+                    strerror(errno));
+            return -1;
         }
+        for (i = 0; i < n_events; i++) {
```
Comment (ShixiongQi) [inline src/gateway.c:274]: Add an empty line before this line.

File: src/gateway.c [hunk 6]
```diff
@@ -207,81 +222,116 @@ static int rpc_server(void) {
...
+        return -1;
     }
+    return 0;
+}
+static int rpc_server_receive(int epfd) {
+    int ret;
+    int n_events;
+    int i;
+    int sockfd_c;
+    struct epoll_event event[N_EVENTS_MAX];
```
Comment (ShixiongQi) [inline src/gateway.c]: remove this line.

File: src/gateway.c [hunk 7]
```diff
@@ -468,6 +518,8 @@ static int conn_write(int *sockfd)
     }
 
     // Inter-node Communication
+    // The ebpf will direct the unkonwn fn_id to the gateway
```
Comment (ShixiongQi) [inline src/gateway.c]: Remove these comments.

File: src/gateway.c [hunk 8]
```diff
@@ -499,15 +551,25 @@ static int conn_write(int *sockfd)
 
         return 1;
     }
-
+    // if the fn_id is 0, it's non direct functino routing
+    // Then just get the next hop then send back locally
```
Comment (ShixiongQi) [inline src/gateway.c]: Remove these comments.

File: src/gateway.c [hunk 9]
```diff
@@ -499,15 +551,25 @@ static int conn_write(int *sockfd)
...
     txn->hop_count++;
 
     log_debug("Next hop is %u", cfg->route[txn->route_id].hop[txn->hop_count]);
+    // fix the route problem without DFR (direct function routing)
+    // fn 1, 2, 3 on node 0, fn 4 on node 1
+    // {
+    //     id = 2;
+    //     name = "Route 2";
+
+    //     hops = [1, 0, 2, 0, 3, 0, 4, 1];
```
Comment (ShixiongQi) [inline src/gateway.c]: Remove these comments.

File: src/gateway.c [hunk 10]
```diff
@@ -603,7 +667,17 @@ static int server_init(struct server_vars *sv)
         return -1;
     }
 
-    ret = pthread_create(&rpc_svr_thread, NULL, &rpc_server_thread, NULL);
+    // create epoll fd for the rpc server
+    epfd = epoll_create1(0);
```
Comment (ShixiongQi) [inline src/gateway.c]: Change `epfd` to `rpc_svr_epfd`

**Review by ShixiongQi — CHANGES_REQUESTED** [review]

**Comment by sungyu-chang** [comment]
Updated by force update to latest commit

**Review by ShixiongQi — APPROVED** [review]
lgtm

**Pull request merged**